File: code/Python/fake_crypto_feed/src/fake_crypto_feed/sender.py
import asyncio
import datetime
import functools
import json
import random
import logging

import psutil
from prometheus_client import start_http_server, Gauge, Counter
import websockets
from websockets import WebSocketServer
import multiprocessing
from multiprocessing.synchronize import Event

# Initialize Prometheus metrics
cpu_usage = Gauge('Python_sender_cpu_usage_percent', 'CPU usage percentage of the data sender')
ram_usage = Gauge('Python_sender_ram_usage_mb', 'RAM usage in MB of the data sender')
io_read = Gauge('sender_io_read_bytes', 'IO read bytes per second of the data sender')
io_write = Gauge('sender_io_write_bytes', 'IO write bytes per second of the data sender')
sent_messages = Counter('Python_sender_sent_messages_total', 'Total number of messages sent by the data sender')
error_count = Counter('Python_sender_error_count_total', 'Total number of errors encountered by the data sender')

async def collect_metrics() -> None:
    process = psutil.Process()

    # Initialize CPU measurement
    process.cpu_percent(interval=None)
    prev_io = process.io_counters()

    while True:
        
        # Get CPU usage
        cpu = process.cpu_percent(interval=None)
        cpu_usage.set(cpu)

        # Get RAM usage in MB
        ram = process.memory_info().rss / (1024 * 1024)
        ram_usage.set(ram)

        # Get I/O counters
        io_counters = process.io_counters()
        io_r = io_counters.read_bytes - prev_io.read_bytes
        io_w = io_counters.write_bytes - prev_io.write_bytes
        prev_io = io_counters

        io_read.set(io_r)
        io_write.set(io_w)

        await asyncio.sleep(1)  # Update every 5 seconds

async def price_feed(websocket: WebSocketServer, _path: str, interval: float) -> None:
    bid_price = random.uniform(10000, 20000)
    ask_price = bid_price + random.uniform(5, 15)

    while True:
        try:
            jitter = random.uniform(-0.1, 0.1)  # Jitter between -10% to +10%
            actual_interval = interval * (1 + jitter)
            await asyncio.sleep(actual_interval)

            # Determine price change direction
            direction = random.choice([-1, 1])

            # Generate percentage change between 1% and 3%
            change_percentage = random.uniform(0.01, 0.03)

            # Update bid_price and ask_price
            bid_change = bid_price * change_percentage * direction
            ask_change = ask_price * (change_percentage + random.uniform(-0.005, 0.005)) * direction

            bid_price += bid_change
            ask_price += ask_change

            if ask_price <= bid_price:
                ask_price = bid_price + random.uniform(5, 15)

            data = {
                "timestamp": datetime.datetime.utcnow().isoformat() + "Z",
                "bid_price": round(bid_price, 2),
                "ask_price": round(ask_price, 2)
            }

            await websocket.send(json.dumps(data))
            logging.info(data)
            sent_messages.inc()

        except Exception as e:
            logging.error(f"Error: {e}")
            error_count.inc()


async def start_server(start_event: Event, wb_port: int=8765, interval: float = 0.01) -> None:
    handler = functools.partial(price_feed, interval=interval)
    server = await websockets.serve(handler, 'localhost', wb_port)
    start_event.set()  # Signal that the server is ready
    await server.wait_closed()
    

def run_sender(start_event: Event, prometheus_port: int=8000, wb_port: int=8765) -> None:
    # Start Prometheus metrics server on port 8000
    start_http_server(prometheus_port)
    
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    loop.create_task(collect_metrics())

    # Signal that the server is starting
    # start_event.set()

    loop.run_until_complete(start_server(start_event, wb_port))


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
    run_sender(multiprocessing.Event())

File: archive/code_backup/Go/fake_crypto_feed/sender/sender.go
package sender

import (
	"context"
	"encoding/json"
	"log"
	"math"
	"math/rand"
	"net/http"
	"os"
	"sync"
	"time"

	"github.com/gorilla/websocket"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/shirou/gopsutil/process"
)

var (
	bytesSent = prometheus.NewCounter(prometheus.CounterOpts{
		Name: "Go_sender_bytes_sent_total",
		Help: "Total bytes sent over the network",
	})
	messagesSent = prometheus.NewCounter(prometheus.CounterOpts{
		Name: "Go_sender_messages_sent_total",
		Help: "Total number of messages sent",
	})
	errorsEncountered = prometheus.NewCounter(prometheus.CounterOpts{
		Name: "Go_sender_errors_total",
		Help: "Total number of errors encountered",
	})
	cpuUsage = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "Go_sender_cpu_usage_percent",
		Help: "CPU usage percentage",
	})
	ramUsage = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "Go_sender_ram_usage_mb",
		Help: "RAM usage in MB",
	})
)

func init() {
	prometheus.MustRegister(bytesSent, messagesSent, errorsEncountered, cpuUsage, ramUsage)
}

var upgrader = websocket.Upgrader{}

type PriceData struct {
	Timestamp string `json:"timestamp"`
	Bid float64 `json:"bid"`
	Ask float64 `json:"ask"`
}

func StartServer(ctx context.Context, wg *sync.WaitGroup) error {
	defer wg.Done()

	// Create a new ServeMux for the main server
	mux := http.NewServeMux()
	mux.HandleFunc("/ws", handleConnections)

	// Create the main server on port 8081
	srv := &http.Server{
		Addr:	":8081",
		Handler: mux,
	}

	// Create a new ServeMux for the metrics server
	metricsMux := http.NewServeMux()
	metricsMux.Handle("/metrics", promhttp.Handler())

	metricsServer := &http.Server{
		Addr:	":9000",
		Handler: metricsMux,
	}

	// Start the server in a goroutine
	internalWG := sync.WaitGroup{}

	internalWG.Add(1)
	go func() {
		defer internalWG.Done()
		log.Println("WebSocket server started on : 8081")
		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Println("WebSocket server error:", err)
		}
	}()

	// Start the metrics server in a separate goroutine
	internalWG.Add(1)
	go func() {
		defer internalWG.Done()
		log.Println("Metrics server started on : 9000")
		if err := metricsServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Println("Metrics server error:", err)
		}

	}()

	// Start updating usage metrics in a goroutine
    internalWG.Add(1)
    go func() {
        defer internalWG.Done()
        updateUsageMetrics(ctx)
    }()

	// Start the shutdown handler in a separate goroutine
    internalWG.Add(1)
	go func() {
		defer internalWG.Done()
		<-ctx.Done()
		log.Println("Shutting down servers...")
		
		// Create a context with a timeout for the shutdown
		shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()

		// Shutdown the main server
		if err := srv.Shutdown(shutdownCtx); err != nil {
			log.Println("Main server shutdown error:", err)
		} else {
			log.Println("Main server gracefully stopped")
		}

		// Shutdown the metrics server
		if err := metricsServer.Shutdown(shutdownCtx); err != nil {
			log.Println("Metrics server shutdown error:", err)
		} else {
			log.Println("Metrics server gracefully stopped")
		}
	}()

	// Wait for all internal goroutines to finish
    internalWG.Wait()

	return nil
}

func handleConnections(w http.ResponseWriter, r *http.Request) {
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Println("Upgrade error:", err)
		errorsEncountered.Inc()
		return
	}
	defer conn.Close()

	// Create a data generator
	dataGenerator := generateFakeDataGenerator()

	for {
		select {
		case <-r.Context().Done():
			return
		default:
			// Get data and interval
            data, interval := dataGenerator()

			message, err := json.Marshal(data)
			if err != nil {
				log.Println("JSON marshal error:", err)
				errorsEncountered.Inc()
				continue
			}
			err = conn.WriteMessage(websocket.TextMessage, message)
			if err != nil {
				log.Println("Write message error:", err)
				errorsEncountered.Inc()
				return
			}
			bytesSent.Add(float64(len(message)))
			messagesSent.Inc()

			// Sleep for the calculated interval
            time.Sleep(interval)
		}
	}
}

// Function to generate a data generator maintaining state
func generateFakeDataGenerator() func() (PriceData, time.Duration) {
    // Initialize bidPrice and askPrice
    bidPrice := rand.Float64()*10000 + 10000 // Between 10000 and 20000
    askPrice := bidPrice + rand.Float64()*10 + 5 // Bid + 5 to 15

    return func() (PriceData, time.Duration) {
        // Calculate jitter between -0.1 and 0.1
        jitter := rand.Float64()*0.2 - 0.1
        interval := time.Duration(200*(1+jitter)) * time.Millisecond

        // Choose direction -1 or 1
        directions := []float64{-1, 1}
        direction := directions[rand.Intn(len(directions))]

        // Change percentage between 0.01 and 0.03
        changePercentage := rand.Float64()*0.02 + 0.01

        // Calculate bid and ask changes
        bidChange := bidPrice * changePercentage * direction
        askChange := askPrice * (changePercentage + rand.Float64()*0.01 - 0.005) * direction

        // Update bidPrice and askPrice
        bidPrice += bidChange
        askPrice += askChange

        // Ensure askPrice > bidPrice
        if askPrice <= bidPrice {
            askPrice = bidPrice + rand.Float64()*10 + 5 // Bid + 5 to 15
        }

        // Round prices to two decimal places
        bidPrice = math.Round(bidPrice*100) / 100
        askPrice = math.Round(askPrice*100) / 100

        // Generate PriceData with timestamp
        data := PriceData{
            Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
            Bid:       bidPrice,
            Ask:       askPrice,
        }

        return data, interval
    }
}


func updateUsageMetrics(ctx context.Context) {
	proc, err := process.NewProcess(int32(os.Getpid()))
	if err != nil {
		log.Println("Process error:", err)
		return
	}
	ticker := time.NewTicker(2 * time.Second) // update every 2 seconds
	defer ticker.Stop() // prevent memory leak

	for {
		select {
		case <-ctx.Done():
			return  // gracefully exit
		case <-ticker.C:
			cpu, err := proc.CPUPercent()
			if err == nil {
				cpuUsage.Set(cpu)
			}
			memInfo, err := proc.MemoryInfo()
			if err == nil {
				ramUsage.Set(float64(memInfo.RSS) / 1024 / 1024)
			}
		}
	}
}
